wgpu-core/src/command/transfer.rs

```diff
@@ -374,60 +376,110 @@ pub(crate) fn validate_texture_copy_range(
+    if !immediate_inits.is_empty() {
+        let cmd_buf_raw = cmd_buf.encoder.open();
+        for init in immediate_inits {
+            clear_texture(
+                &init.texture,
+                texture,
+                TextureInitRange {
+                    mip_range: init.mip_level..(init.mip_level + 1),
+                    layer_range: init.layer..(init.layer + 1),
+                },
+                cmd_buf_raw,
+                &mut cmd_buf.trackers.textures,
+                device,
+            )
+            .unwrap();
+        }
     }
 }
 
+// Ensures the source texture of a transfer is in the right initialization state and records the state for after the transfer operation.
 fn handle_src_texture_init<A: hal::Api>(
     cmd_buf: &mut CommandBuffer<A>,
     device: &Device<A>,
     source: &ImageCopyTexture,
-    src_base: &hal::TextureCopyBase,
     copy_size: &Extent3d,
     texture_guard: &Storage<Texture<A>, TextureId>,
-) {
-    let immediate_src_init = cmd_buf.texture_memory_actions.register_init_action(
-        &TextureInitTrackerAction {
-            id: source.texture,
-            range: TextureInitRange {
-                mip_range: src_base.mip_level..src_base.mip_level + 1,
-                layer_range: src_base.origin.z
-                    ..(src_base.origin.z + copy_size.depth_or_array_layers),
-            },
-            kind: MemoryInitKind::NeedsInitializedMemory,
-        },
+) -> Result<(), TransferError> {
+    let texture = texture_guard
+        .get(source.texture)
+        .map_err(|_| TransferError::InvalidTexture(source.texture))?;
+
+    handle_texture_init(
+        MemoryInitKind::NeedsInitializedMemory,
+        cmd_buf,
+        device,
+        source,
+        copy_size,
         texture_guard,
+        texture,
     );
-    if !immediate_src_init.is_empty() {
-        let cmd_buf_raw = cmd_buf.encoder.open();
-        fixup_discarded_surfaces(
-            immediate_src_init.into_iter(),
-            cmd_buf_raw,
-            texture_guard,
-            &mut cmd_buf.trackers.textures,
-            device,
-        );
-    }
+    Ok(())
+}
+
+// Ensures the destination texture of a transfer is in the right initialization state and records the state for after the transfer operation.
+fn handle_dst_texture_init<A: hal::Api>(
+    cmd_buf: &mut CommandBuffer<A>,
+    device: &Device<A>,
+    destination: &ImageCopyTexture,
+    copy_size: &Extent3d,
+    texture_guard: &Storage<Texture<A>, TextureId>,
+) -> Result<(), TransferError> {
+    let texture = texture_guard
+        .get(destination.texture)
+        .map_err(|_| TransferError::InvalidTexture(destination.texture))?;
+
+    // Attention: If we don't write full texture subresources, we need to a full clear first since we don't track subrects.
+    // This means that in rare cases even a *destination* texture of a transfer may need an immediate texture init.
+    let dst_init_kind = if copy_size.width == texture.desc.size.width
```

Comment:
This check is incorrect, it doesn't take into account the mip level size. Instead, it should use `texture.desc.mip_level_size(destination.mip_level)`